libvast/src/synopsis.cpp

```diff
@@ -0,0 +1,119 @@
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#include "vast/synopsis.hpp"
+
+#include <caf/actor_system.hpp>
+#include <caf/runtime_settings_map.hpp>
+
+#include "vast/error.hpp"
+#include "vast/min_max_synopsis.hpp"
+
+#include "vast/detail/overload.hpp"
+
+namespace vast {
+
+synopsis::synopsis(vast::type x) : type_{std::move(x)} {
+  // nop
+}
+
+synopsis::~synopsis() {
+  // nop
+}
+
+const vast::type& synopsis::type() const {
+  return type_;
+}
+
+caf::error inspect(caf::serializer& sink, synopsis_ptr& ptr) {
+  if (!ptr) {
+    type dummy;
+    return sink(dummy);
+  }
+  return caf::error::eval([&] { return sink(ptr->type()); },
+                          [&] { return ptr->serialize(sink); });
+}
+
+caf::error inspect(caf::deserializer& source, synopsis_ptr& ptr) {
+  type t;
+  auto err = source(t);
+  if (err)
+    return err;
+  // Only default-constructed synopses have an empty type.
+  if (!t) {
+    ptr.reset();
+    return caf::none;
+  }
+  if (source.context() != nullptr) {
+    auto factory = get_synopsis_factory_fun(source.context()->system());
+    ptr = factory ? factory(std::move(t)) : make_synopsis(std::move(t));
+  } else {
+    ptr = make_synopsis(std::move(t));
+  }
+  return ptr->deserialize(source);
+}
+
+namespace {
+
+class timestamp_synopsis : public min_max_synopsis<timestamp> {
+public:
+  timestamp_synopsis(vast::type x) 
+    : min_max_synopsis<timestamp>{std::move(x), timestamp::max(),
+                                  timestamp::min()} {
+    // nop
+  }
+};
+
+constexpr auto synopis_factory_fun_atom = caf::atom("SYNOPSIS_F");
+constexpr auto synopis_factory_tag_atom = caf::atom("SYNOPSIS_T");
+
+} // namespace <anonymous>
+
+synopsis_ptr make_synopsis(type x) {
+  return caf::visit(detail::overload(
+    [&](const timestamp_type&) -> synopsis_ptr {
+      return caf::make_counted<timestamp_synopsis>(std::move(x));
+    },
+    [](const auto&) -> synopsis_ptr {
+      return nullptr;
+    }), x);
+}
+
+// TODO: we should find a way to associate a key-value pair with the
+// runtime-settings map, with key being an atom and value a function pointer.
+// Right now, it's a brittle setup where the user must manage two keys.
+
+synopsis_factory get_synopsis_factory_fun(caf::actor_system& sys) {
+  using generic_fun = caf::runtime_settings_map::generic_function_pointer;
+  auto val = sys.runtime_settings().get(synopis_factory_fun_atom);
+  if (auto fun = caf::get_if<generic_fun>(&val))
+    return reinterpret_cast<synopsis_ptr (*)(type)>(*fun);
+  return {};
+}
+
+caf::atom_value get_synopsis_factory_tag(caf::actor_system& sys) {
+  auto val = sys.runtime_settings().get(synopis_factory_tag_atom);
+  if (auto x = caf::get_if<caf::atom_value>(&val))
+    return *x;
+  return {};
+}
+
+void set_synopsis_factory(caf::actor_system& sys,
```

Comment:
I'm missing a usage example of this `set_synopsis_factory` stuff. As far as I understand it now, you need to call the overload provided here to register the factory and tag, and then call the overload provided in `meta_index` to copy a pointer to the factory in there. That pointer is used for serialization, but for deserialization the registry is used directly. This seems rather error prone.

Reply:
Per our discussion in the office today, we agreed to keep the current attempt as is and then @Neverlord will take over and perform the adaptations to mirror the factory implementation of table slices.